src/buildHTML.js

```diff
@@ -88,146 +64,173 @@ export const buildExpression = function(
-            // if current non-space node is left dummy span, add a glue before
-            // first real non-space node
-            if (j === 0) {
-                groups.pop();
-                i--;
-            }
+        // Ignore explicit spaces (e.g., \;, \,) when determining what implicit
+        // spacing should go between atoms of different classes
+        if (node.classes[0] === "mspace") {
+            continue;
+        }
 
-            // Get the type of the current non-space node.  If it's a document
-            // fragment, get the type of the rightmost node in the fragment.
-            const left = getTypeOfDomTree(nonSpaces[j], "right");
-
-            // Get the type of the next non-space node.  If it's a document
-            // fragment, get the type of the leftmost node in the fragment.
-            const right = getTypeOfDomTree(nonSpaces[j + 1], "left");
-
-            // We use buildExpression inside of sizingGroup, but it returns a
-            // document fragment of elements.  sizingGroup sets `isRealGroup`
-            // to false to avoid processing spans multiple times.
-            if (left && right && isRealGroup) {
-                const nonSpacesJp1: HtmlDomNode = assert(nonSpaces[j + 1]);
-                const space = isLeftTight(nonSpacesJp1)
-                    ? tightSpacings[left][right]
-                    : spacings[left][right];
-
-                if (space) {
-                    let glueOptions = options;
-
-                    if (expression.length === 1) {
-                        const node =
-                            checkNodeType(expression[0], "sizing") ||
-                            checkNodeType(expression[0], "styling");
-                        if (!node) {
-                            // No match.
-                        } else if (node.type === "sizing") {
-                            glueOptions = options.havingSize(node.size);
-                        } else if (node.type === "styling") {
-                            glueOptions = options.havingStyle(styleMap[node.style]);
-                        }
-                    }
-
-                    groups.push(buildCommon.makeGlue(space, glueOptions));
-                }
+        const result = callback(node, prev.node);
+        if (result) {
+            if (prev.insertAfter) {
+                prev.insertAfter(result);
+            } else { // insert at front
+                nodes.unshift(result);
+                i++;
             }
-            j++;
         }
+
+        prev.node = node;
+        prev.insertAfter = (index => n => {
+            nodes.splice(index + 1, 0, n);
+            i++;
+        })(i);
     }
+    if (next) {
+        nodes.pop();
+    }
+};
 
-    return groups;
+// Check if given node is a partial group, i.e., does not affect spacing around,
+// like fragment.
+const checkPartialGroup = function(
+    node: HtmlDomNode,
+): ?(DocumentFragment<HtmlDomNode> | Anchor) {
```

Comment:
Maybe we should define `type PartialGroup = DocumentFragment<HtmlDomNode> | Anchor`.